### DISCUSSION TITLE
[Help] How can I get title knowing note's id with LSP?

### DISCUSSION THREAD

**Discussion by kirasok** [zk-org/zk-nvim#191] [Help] [answered]
### Verify

- [X] I searched the existing discussions for help

### How can we help you?

I am trying to solve issue #41 using [render-markdown.nvim](https://github.com/MeanderingProgrammer/render-markdown.nvim). Currently, it supports concealing [wikilinks](https://github.com/MeanderingProgrammer/render-markdown.nvim/wiki/Links). But by default ``[[id]]`` becomes ``󰌹 id``. I want to replace id with title. 

I found a way to get title by using cli `zk list $(find . -type f -name "id*.md") --format=json --quiet | jq '.[0].title'` but that doesn't seem to be the best way because lsp already provides diagnostics for wikilink titles

Is there way to get note's title knowing it's id by using lsp?

**Comment by tjex** [comment]
In which context are you wanting to "get" the note's title?
Within your config, within neovim, from zk cli?

The notes title is exposed in some way as mentioned in that same issue: https://github.com/zk-org/zk-nvim/issues/41#issuecomment-1045981698

But I'm guessing that's not working for you / not in the right direction.

But yeah, unfortunately I'm not completely understanding what you're needing, where you need it and how it should be implemented for render-markdown.nvim (I'm not familiar with this plugin).

**Reply by kirasok** [reply, marked as answer]
I wanted to conceal wikilinks, and neovim doesn't conceal them. I found [render-markdown.nvim](https://github.com/MeanderingProgrammer/render-markdown.nvim) plugin but it only conceals wikilinks with custom title, which is not the case [for me](https://github.com/zk-org/zk-nvim/discussions/191#discussioncomment-10867807). So I decided to make a fork that adds support of getting title from language server.

For this, I was looking how to get note's title by knowing note's id within neovim. I knew neovim gets diagnostics from `zk lsp` but I didn't know what function I should use. 

I haven't found what I was looking for but I managed to get diagnostics with `vim.diagnostic.get()`. Then within [markdown handler of render-markdown.nvim](https://github.com/kirasok/render-markdown.nvim/blob/cfaa6b8eb0ceee544bc13a33ed52a5b5285f6096/lua/render-markdown/handler/markdown_inline.lua#L169) I compared node's position with provided diagnostics, and then displayed the message.

```markdown
# index

- [[rigo|first note]]
- [second note](zzl2)
- [[330x]]

```

<details>
<summary>nvim-treesitter conceal</summary>
<img src="https://github.com/user-attachments/assets/6f41cbeb-b7d5-4719-8644-ce028cfb58c1"></img>
</details>

<details>
<summary>render-markdown.nvim</summary>
<img src="https://github.com/user-attachments/assets/71cfd4b2-865b-4165-a4cb-9cc9cd0d4302"></img>
</details>

<details>
<summary>my fork</summary>
<img src="https://github.com/user-attachments/assets/32952213-2c46-46b4-a9b9-55d96f6cddd7"></img>
</details>

**Reply by tjex** [reply]
Oh wow. Nice work. It would be awesome if you could make a tiny post about this in the "share" section of the discussions.

No doubt others would love to use this!

**Comment by tjex** [comment]
@kirasok Does it not suffice to just use the notes title as the link title? Which is default behaviour.

Or use a template for it.

https://zk-org.github.io/zk/notes/note-format.html#markdown

**Reply by kirasok** [reply]
Yes, setting `link-format` to `markdown` or `[[{{metadata.id}}|{{title}}]]` works for new links. But I used `[[{filename}]]` since creating notebook and it would take too much time to convert all the links in all notes I have.
